File: leetcode_solved/leetcode_1143_Longest_Common_Subsequence.cpp
class Solution {
public:
	/**
	 * 思路：此题要求两个字符串的最长公共子序列，返回此公共子序列的长度。
	 * 容易和另外一个问题搞混淆，这里不是求最长公共子串，子串里各个字符是连续出现在原字符串，子序列不需要连续
	 * 采用查表法。具体分析过程看下面帖子
	 *
	 * @see 
	 */
    int longestCommonSubsequence(string &a, string &b) {
      vector<vector<short>> m(a.size() + 1, vector<short>(b.size() + 1));
      for (auto i = 1; i <= a.size(); ++i)
        for (auto j = 1; j <= b.size(); ++j)
          if (a[i - 1] == b[j - 1]) m[i][j] = m[i - 1][j - 1] + 1;
          else m[i][j] = max(m[i - 1][j], m[i][j - 1]);
      return m[a.size()][b.size()];
    }
};

File: leetcode_solved/leetcode_1143_Longest_Common_Subsequence.java
// AC: Runtime: 16 ms, faster than 66.93% of Java online submissions for Longest Common Subsequence.
// Memory Usage: 46.6 MB, less than 12.77% of Java online submissions for Longest Common Subsequence.
// dynamic programming
// T:O(m * n), S:O(m * n)
// 
class Solution {
    public int longestCommonSubsequence(String text1, String text2) {
        int len1 = text1.length(), len2 = text2.length();
        int[][] dp = new int[len1 + 1][len2 + 1];
        for (int i = 0; i < len1; i++) {
            for (int j = 0; j < len2; j++) {
                dp[i + 1][j + 1] = text1.charAt(i) == text2.charAt(j) ? dp[i][j] + 1 : Math.max(dp[i][j + 1], dp[i + 1][j]);
            }
        }

        return dp[len1][len2];
    }
}